sipp/src/lib.rs

```diff
@@ -170,39 +187,28 @@ impl<E: PairingEngine, D: Digest> SIPP<E, D> {
 }
 
 /// Compute the product of pairings of `r_i * a_i` and `b_i`.
-pub fn product_of_pairings_with_coeffs<E: PairingEngine>(
+pub fn product_of_pairings_with_coeffs<E: Pairing>(
     a: &[E::G1Affine],
     b: &[E::G2Affine],
-    r: &[E::Fr],
-) -> E::Fqk {
+    r: &[E::ScalarField],
+) -> PairingOutput<E> {
     let a = a
         .into_par_iter()
         .zip(r)
-        .map(|(a, r)| a.mul(*r))
-        .collect::<Vec<_>>();
-    let a = E::G1Projective::batch_normalization_into_affine(&a);
-    let elements = a
-        .par_iter()
-        .zip(b)
-        .map(|(a, b)| (E::G1Prepared::from(*a), E::G2Prepared::from(*b)))
+        .map(|(&a, r)| a * r)
         .collect::<Vec<_>>();
-    let num_chunks = elements.len() / rayon::current_num_threads();
-    let num_chunks = if num_chunks == 0 {
-        elements.len()
-    } else {
-        num_chunks
-    };
-    let ml_result = elements
-        .par_chunks(num_chunks)
-        .map(E::miller_loop)
-        .product();
-    E::final_exponentiation(&ml_result).unwrap()
+    let a = E::G1::normalize_batch(&a);
+
+    let a = a.par_iter().map(E::G1Prepared::from).collect::<Vec<_>>();
+    let b = b.par_iter().map(E::G2Prepared::from).collect::<Vec<_>>();
+
+    E::multi_pairing(a, b)
```

Comment:
`multi_pairing` doesn't parallelize internally, so it might still be worthwhile to do that manually here.

Reply:
I was starting to add some support in `algebra`, but it looks like the impls do parallelize internally. [BLS](https://github.com/arkworks-rs/algebra/blob/d106993409c85cb3eaf2ad0c45c7076089c10e7d/ec/src/models/bls12/mod.rs#L64), [MNT4](https://github.com/arkworks-rs/algebra/blob/d106993409c85cb3eaf2ad0c45c7076089c10e7d/ec/src/models/mnt4/mod.rs#L55), [MNT6](https://github.com/arkworks-rs/algebra/blob/d106993409c85cb3eaf2ad0c45c7076089c10e7d/ec/src/models/mnt6/mod.rs#L56), [BW6](https://github.com/arkworks-rs/algebra/blob/d106993409c85cb3eaf2ad0c45c7076089c10e7d/ec/src/models/bw6/mod.rs#L100), and [BN256](https://github.com/arkworks-rs/algebra/blob/d106993409c85cb3eaf2ad0c45c7076089c10e7d/ec/src/models/bn/mod.rs#L66) (not fully)

Reply:
Copied `cfg_multi_pairing` functionality here as well. Can't use the export bc `sipp` doesn't have anything in this crate as a dep